src/ring_theory/dedekind_domain/finite_adele_ring.lean

```diff
@@ -0,0 +1,292 @@
+/-
+Copyright (c) 2022 María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: María Inés de Frutos-Fernández
+-/
+import ring_theory.dedekind_domain.adic_valuation
+import topology.algebra.uniform_ring
+
+
+/-!
+# The finite adèle ring of a Dedekind domain
+We define the ring of finite adèles of a Dedekind domain `R`.
+
+## Main definitions
+- `prod_completions_integers` : product of `adic_completion_integers`, where `v` runs over all
+   maximal ideals of `R`.
+- `prod_completions` : the product of `adic_completion`, where `v` runs over all maximal ideals
+  of `R`.
+- `finite_adele_ring` : The finite adèle ring of `R`, defined as the restricted product
+  `Π'_v K_v`.
+
+## Implementation notes
+We are only interested on Dedekind domains of Krull dimension 1 (i.e., not fields). If `R` is a
+field, its finite adèle ring is just defined to be the trivial ring.
+
+## References
+* [J.W.S. Cassels, A. Frölich, *Algebraic Number Theory*][cassels1967algebraic]
+
+## Tags
+finite adèle ring, dedekind domain
+-/
+
+noncomputable theory
+open function set is_dedekind_domain is_dedekind_domain.height_one_spectrum
+
+variables (R : Type) (K : Type) [comm_ring R] [is_domain R] [is_dedekind_domain R] [field K]
+  [algebra R K] [is_fraction_ring R K] (v : height_one_spectrum R)
+
+/-- The product of all `adic_completion_integers`, where `v` runs over the maximal ideals of `R`. -/
+def prod_completions_integers := (Π (v : height_one_spectrum R), v.adic_completion_integers K)
+local notation `R_hat` := prod_completions_integers
+
+instance : comm_ring (R_hat R K) := pi.comm_ring
+
+instance : topological_space (R_hat R K) := Pi.topological_space
+
+instance prod_completions_integers.inhabited : inhabited (prod_completions_integers R K) := ⟨0⟩
+
+/-- The product of all `adic_completion`, where `v` runs over the maximal ideals of `R`. -/
+def prod_completions := (Π (v : height_one_spectrum R), v.adic_completion K)
```

Comment:
OK, I'll try to come up with a better name; let me think about it for a bit.

Reply:
How about you make a `dedekind_domain` namespace and put some of this stuff in there? I think there is an argument for "filing it away" in such a namespace, because these constructions all depend on `height_one_spectrum` which needs `is_dedekind_domain`. I guess one issue with `prod_completions` is that it does not consider the completions at the infinite places. Is `dedekind_domain.prod_finite_completions` even worse? Or `dedekind_domain.prod_adic_completions`?

Reply:
I like the `dedekind_domain.prod_adic_completions` best. I have moved everything in the file to the `dedekind_domain` namespace.